localstack/services/s3/models_native.py

```diff
@@ -0,0 +1,486 @@
+    is_current: bool
+
+    def __init__(self, key: ObjectKey, version_id: ObjectVersionId):
+        self.key = key
+        self.version_id = version_id
+        self.last_modified = datetime.utcnow()
+        self.is_current = True
+
+
+# TODO: use SynchronizedDefaultDict to prevent updates during iteration?
+class KeyStore:
+    """
+    Object representing an S3 Un-versioned Bucket's Key Store. An object is mapped by a key, and you can simply
+    retrieve the object from that key.
+    """
+
+    def __init__(self):
+        self._store = {}
+
+    def get(self, object_key: ObjectKey) -> S3Object | None:
+        return self._store.get(object_key)
+
+    def set(self, object_key: ObjectKey, s3_object: S3Object):
+        self._store[object_key] = s3_object
+
+    def pop(self, object_key: ObjectKey, default=None) -> S3Object | None:
+        return self._store.pop(object_key, default)
+
+    def values(self, *_, **__) -> list[S3Object | S3DeleteMarker]:
+        return [value for value in self._store.values()]
+
+    def is_empty(self) -> bool:
+        return not self._store
+
+
+class VersionedKeyStore:
+    """
+    Object representing an S3 Versioned Bucket's Key Store. An object is mapped by a key, and adding an object to the
+    same key will create a new version of it. When deleting the object, a S3DeleteMarker is created and put on top
+    of the version stack, to signal the object has been "deleted".
+    This object allows easy retrieval and saving of new object versions.
+    See: https://docs.aws.amazon.com/AmazonS3/latest/userguide/versioning-workflows.html
+    """
+
+    def __init__(self):
+        self._store = defaultdict(dict)
+
+    def get(
+        self, object_key: ObjectKey, version_id: ObjectVersionId = None
+    ) -> S3Object | S3DeleteMarker | None:
+        """
+        :param object_key: the key of the Object we need to retrieve
+        :param version_id: Optional, if not specified, return the current version (last one inserted)
+        :return: an S3Object or S3DeleteMarker
+        """
+        if not version_id and (versions := self._store.get(object_key)):
+            for version_id in reversed(versions):
+                return versions.get(version_id)
+
+        return self._store.get(object_key, {}).get(version_id)
+
+    def set(self, object_key: ObjectKey, s3_object: S3Object | S3DeleteMarker):
+        """
+        Set an S3 object, using its already set VersionId.
+        If the bucket versioning is `Enabled`, then we're just inserting a new Version.
+        If the bucket versioning is `Suspended`, the current object version will be set to `null`, so if setting a new
+        object at the same key, we will override it at the `null` versionId entry.
+        :param object_key: the key of the Object we are setting
+        :param s3_object: the S3 object or S3DeleteMarker to set
+        :return: None
+        """
+        self._store[object_key][s3_object.version_id] = S3Object
```

Comment:
```suggestion
        self._store[object_key][s3_object.version_id] = s3_object
```
whoops, always the same issue with versioning and couldn't be tested. Will be part of the versioning PR 👍